doc/design/state-machine.md

```diff
@@ -0,0 +1,185 @@
+# HiveD State Machines
+
+This document presents the state machines of HiveD, and explains the life cycles of jobs and resources in HiveD.
+We will describe the state machines of our scheduling unit, Affinity Group, and that of our resource unit, Cell, respectively.
+
+## Affinity Group (AG) State Machine
+
+An affinity group (AG) is a set of gang-scheduled pods. It is the basic scheduling unit in HiveD. The figure below shows the state machine of an affinity group.
+
+Note that the AG state machine has interactions with the cell state machine (elaborated later). In our design, AGs influence each other only via cell: an event of an AG may trigger an event for one of its cells, and the influenced cell may trigger another event for another AG related to the same cell.
+Therefore, the state machines of multiple AGs are effectively bridged by the state machines of their overlapping cells.
+
+![AG](AG-state-machine.png)
+
+### States
+
+__Pending__: the AG is waiting to be scheduled.
+
+__Preempting__: the AG is allocated resources but is waiting for the completion of preemptions of other AGs (only after that can it start to run).
+
+__Allocated__: the AG is allocated resources and is running normally.
+
+__Being preempted__: the AG is being preempted by other AGs (the preemption is still ongoing).
+
+__Deleted__: the AG is fully deleted.
+
+Common AG life cycles:
+
+Pending -> Allocated -> Deleted (no preemption involved)
+
+Pending -> Preempting -> Allocated -> Being preempted -> Deleted (preemption involved)
+
+### Events
+
+__e<sub>c</sub>__:
+
+Condition: scheduler crashes and restarts.
+
+Operation: none.
+
+Note that only Pending, Allocated, and Deleted are persistent, thus they are the recovery points of HiveD. While other AG states (Preempting, Being preempted) are volatile, so they will transition to others after e<sub>c</sub>. Also note that Allocated state includes updating pod annotation, pod binding, and pod running. We assume once pod annotation has been updated, the pod binding and pod running are handled by K8s. 
+
+__e<sub>0</sub>__:
+
+Condition: all cells allocated to the AG are in Free or Reserved state (we assume the cell allocation algorithm has ensured that this AG’s priority is higher than that every Reserved cell, i.e., that of the Preempting AG).
+
+Operation: set the cells to Used; set the Preempting AG on the Reserved cells to Pending.
+
+__e<sub>1</sub>__:
+
+Condition: there is at least one cell allocated to the AG in Used or Reserving state (we assume the cell allocation algorithm has ensured that this AG’s priority is higher than that of every Used and Reserving cell).
+
+Operation: set all the Used cells to Reserving state (e<sub>2</sub> in cell state machine); overwrite the Reserved (e<sub>6</sub> in cell state machine) and Reserving cells (e<sub>3</sub> in cell state machine); set the Free cell to Reserved (e<sub>5</sub> in cell state machine).
```

Comment:
For this maybe below text is more clear (using "->"):
For all cells that needs to be preempt:
Cells Used by others -> Cells Reserving by this (e<sub>2</sub> in cell state machine);
Cells Reserving/Reserved by others -> Cells Reserving/Reserved by this (e<sub>3</sub> e<sub>6</sub> in cell state machine); 
Free Cells -> Cells Reserved by this (e<sub>5</sub> in cell state machine).

Reply:
Also for other places?